server/datastore/mysql/software.go

```diff
@@ -314,12 +314,27 @@ func selectSoftwareSQL(hostID *uint, opts fleet.SoftwareListOptions) (string, []
 	}
 
 	if opts.WithHostCounts {
+		subSelectCounts := dialect.From(goqu.I("aggregated_stats").As("shc")).Select(
```

Comment:
Still not ideal, because we cannot create an index on json_value (points for not using `aggregated_stats` for this, I suppose). But this makes it so it's bellow 5 secs in the dashboard, as we are joining an already filtered table here.

Reply:
For my own understanding, what is the optimization being made here exactly? I don't really follow what the change is - instead of joining on `aggregated_stats`, we join on a sub-select on `aggregated_stats`? I struggle to understand what makes this significantly faster (the goqu DSL doesn't help, admittedly).

Reply:
When we sort by host counts, the way the query works is: join all of software with all of aggregated stats, then sort, then limit.

If the aggregated stats table is already sorted and limited, then the join is only against the rows that matter.

Here's the idea illustrated in plain sql:

```
select s.*, shc.json_value as hosts_count from software s join aggregated_stats shc on (s.id=shc.id) where shc.type='software_hosts_count' and shc.json_value > 0 order by hosts_count limit 10;
+--------+----------------+---------+--------+--------------------------+-------------+
| id     | name           | version | source | bundle_identifier        | hosts_count |
+--------+----------------+---------+--------+--------------------------+-------------+
|  65536 | Unique_3732_7  | 1.1.1   | apps   | com.fleetdm.osquery-perf | 1           |
| 131072 | Unique_12854_2 | 1.1.1   | apps   | com.fleetdm.osquery-perf | 1           |
| 196608 | Unique_19405_7 | 1.1.1   | apps   | com.fleetdm.osquery-perf | 1           |
| 262144 | Unique_25961_3 | 1.1.1   | apps   | com.fleetdm.osquery-perf | 1           |
| 327680 | Unique_32514_8 | 1.1.1   | apps   | com.fleetdm.osquery-perf | 1           |
| 393216 | Unique_39071_7 | 1.1.1   | apps   | com.fleetdm.osquery-perf | 1           |
| 458752 | Unique_45671_9 | 1.1.1   | apps   | com.fleetdm.osquery-perf | 1           |
| 524288 | Unique_52257_6 | 1.1.1   | apps   | com.fleetdm.osquery-perf | 1           |
| 589824 | Unique_58688_8 | 1.1.1   | apps   | com.fleetdm.osquery-perf | 1           |
| 655360 | Unique_65224_1 | 1.1.1   | apps   | com.fleetdm.osquery-perf | 1           |
+--------+----------------+---------+--------+--------------------------+-------------+
10 rows in set (29.71 sec)
```

vs

```
select s.*, shc.hosts_count from software s join (select id, json_value as hosts_count from aggregated_stats where type='software_hosts_count' and json_value > 0 order by hosts_count limit 10) shc on (s.id=shc.id);
+--------+----------------+---------+--------+--------------------------+-------------+
| id     | name           | version | source | bundle_identifier        | hosts_count |
+--------+----------------+---------+--------+--------------------------+-------------+
|  65536 | Unique_3732_7  | 1.1.1   | apps   | com.fleetdm.osquery-perf | 1           |
| 131072 | Unique_12854_2 | 1.1.1   | apps   | com.fleetdm.osquery-perf | 1           |
| 196608 | Unique_19405_7 | 1.1.1   | apps   | com.fleetdm.osquery-perf | 1           |
| 262144 | Unique_25961_3 | 1.1.1   | apps   | com.fleetdm.osquery-perf | 1           |
| 327680 | Unique_32514_8 | 1.1.1   | apps   | com.fleetdm.osquery-perf | 1           |
| 393216 | Unique_39071_7 | 1.1.1   | apps   | com.fleetdm.osquery-perf | 1           |
| 458752 | Unique_45671_9 | 1.1.1   | apps   | com.fleetdm.osquery-perf | 1           |
| 524288 | Unique_52257_6 | 1.1.1   | apps   | com.fleetdm.osquery-perf | 1           |
| 589824 | Unique_58688_8 | 1.1.1   | apps   | com.fleetdm.osquery-perf | 1           |
| 655360 | Unique_65224_1 | 1.1.1   | apps   | com.fleetdm.osquery-perf | 1           |
+--------+----------------+---------+--------+--------------------------+-------------+
10 rows in set (3.71 sec)
```

The explains are not particularly illustrating in this case, not sure why. But the resulting times are pretty consistent.

Reply:
Gotcha, seeing the query like that makes it pretty clear, thanks for the great explanation. Also, I think the ORDER BY is **also** applied to the main query in the actual code, right? (It is not in your example, but I don't think the ORDER BY in the sub-select guarantees any ordering on the returned rows).

Reply:
Right, it's applied to both. The order by in the subselect is just to filter the rows we want when we order by hosts count (it also sorts them, but we don't care about that at this point). And then the outer order by is for the result (which can also be over other columns... so this doesn't solve every case, but optimizes the main use cases we have).